internal/impl/gcp/input_bigquery_select.go

```diff
@@ -0,0 +1,273 @@
+
+type bigQuerySelectInput struct {
+	logger *service.Logger
+	config *bigQuerySelectInputConfig
+
+	// Used for overriding client options in tests
+	clientOpts []option.ClientOption
+	client     *bigquery.Client
+
+	shutdownSig *shutdown.Signaller
+
+	// Represents a row iterator that returns query results
+	// The indirection provided by the `bigqueryIterator` interface allows test
+	// code to conveniently create mock iterators
+	iterator bigqueryIterator
+}
+
+func newBigQuerySelectInput(inConf *service.ParsedConfig, logger *service.Logger) (*bigQuerySelectInput, error) {
+	conf, err := bigQuerySelectInputConfigFromParsed(inConf)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse config: %w", err)
+	}
+
+	return &bigQuerySelectInput{
+		logger:      logger,
+		config:      &conf,
+		shutdownSig: shutdown.NewSignaller(),
+	}, nil
+}
+
+func (inp *bigQuerySelectInput) Connect(ctx context.Context) error {
+	client, err := bigquery.NewClient(context.Background(), inp.config.project, inp.clientOpts...)
+	if err != nil {
+		return fmt.Errorf("failed to create bigquery client: %w", err)
+	}
+
+	inp.client = client
+
+	var args []interface{}
+	argsMapping := inp.config.argsMapping
+
+	if argsMapping != nil {
+		rawArgs, err := inp.config.argsMapping.Query(nil)
+		if err != nil {
+			return err
+		}
+
+		checkedArgs, ok := rawArgs.([]interface{})
+		if !ok {
+			return fmt.Errorf("mapping returned non-array result: %T", rawArgs)
+		}
+
+		args = checkedArgs
+	}
+
+	query, err := buildBQQuery(client, &bqQueryBuilderOptions{
+		queryParts: inp.config.queryParts,
+		jobLabels:  inp.config.jobLabels,
+		args:       args,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to build query: %w", err)
+	}
+
+	inp.logger.Debugf("executing query: %s", query.Q)
+
+	jobctx, _ := inp.shutdownSig.CloseAtLeisureCtx(context.Background())
+
+	job, err := query.Run(jobctx)
+	if err != nil {
+		return fmt.Errorf("failed to run query: %w", err)
+	}
+
+	inp.logger.With("job_id", job.ID()).Debug("running bigquery job")
+
+	status, err := job.Wait(jobctx)
+	if err != nil {
+		return fmt.Errorf("failed to wait on job: %w", err)
+	}
+
+	if err := status.Err(); err != nil {
+		return fmt.Errorf("failed to complete job successfully: %w", err)
+	}
+
+	it, err := job.Read(jobctx)
+	if err != nil {
+		return fmt.Errorf("failed to read rows: %w", err)
+	}
+
+	inp.iterator = it
+
+	return nil
+}
+
+func (inp *bigQuerySelectInput) Read(ctx context.Context) (*service.Message, service.AckFunc, error) {
+	if inp.iterator == nil {
+		return nil, nil, fmt.Errorf("query result iterator is not set: %w", service.ErrNotConnected)
+	}
+
+	var row map[string]bigquery.Value
+	err := inp.iterator.Next(&row)
+	if err == iterator.Done {
+		return nil, nil, service.ErrEndOfInput
+	}
+	if err != nil {
+		return nil, nil, err
+	}
+
+	bs, err := json.Marshal(row)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to marshal row to json: %w", err)
+	}
+
+	msg := service.NewMessage(bs)
```

Comment:
I'm not sure why I need to manually marshal a `map[string]bigquery.Value` to json but passing it as is to `msg.SetStructured` resulted in an "unknown" types that could not be traversed in bloblang.